### ISSUE TITLE
Question regarding schema versioning and usage in react-native-navigation (HOC)

### ISSUE THREAD

**Issue by reinvanimschoot** [apollographql/apollo-cache-persist#44]
Hi everyone,

I had a question regarding the schema versioning functionality in combination with react-native-navigation.

At the moment, I have an async function as described in the documentation: 

```
const SCHEMA_VERSION = '1';
const SCHEMA_VERSION_KEY = 'apollo-schema-version';

const setupClient = async () => {
  // server endpoint
  const uri = Platform.OS === 'ios' ? 'http://localhost:4000' : 'http://10.0.2.2:4000';

  const httpLink = createHttpLink({
    uri,
  });

  // authentication
  const authMiddleware = setContext(async () => {
    const token = await AsyncStorage.getItem('token');

    return {
      headers: {
        Authorization: token ? `Bearer ${token}` : '',
      },
    };
  });

  const httpLinkWithAuth = concat(authMiddleware, httpLink);

  // cache
  const cache = new InMemoryCache({
    addTypename: true,
  });

  const persistor = new CachePersistor({
    cache,
    storage: AsyncStorage,
  });

  const currentVersion = await AsyncStorage.getItem(SCHEMA_VERSION_KEY);

  if (currentVersion === SCHEMA_VERSION) {
    // If the current version matches the latest version,
    // we're good to go and can restore the cache.
    await persistor.restore();
  } else {
    // Otherwise, we'll want to purge the outdated persisted cache
    // and mark ourselves as having updated to the latest version.
    await persistor.purge();
    await AsyncStorage.setItem(SCHEMA_VERSION_KEY, SCHEMA_VERSION);
  }

  // local state
  const stateLink = withClientState({ resolvers, cache, defaults });

  return new ApolloClient({
    link: ApolloLink.from([stateLink, httpLinkWithAuth]),
    cache,
  });
};
```

However, when working with react-native-navigation I also need to have an HOC that is used to wrap all the separate screens I declare, like so:

```
export const withProvider = (Component: React.ComponentType) => {
  const WrappedComponent = (props: object) => (
    <ApolloProvider client={client}>
      <Component {...props} />
    </ApolloProvider>
  );

  return WrappedComponent;
};
```

As you can see, it needs the client in there as well. 

So my question is, what is the best way to add the client to the HOC. When doing it synchronously, it didn't pose problems but now that I have a asynchronous function, I'm not sure what I should do.

I could change the code in the HOC to have an asynchronous componentWillMount, where it waits for the client. But I'm not sure if that is optimal since it will do that for all screens while the client needs to be setup asynchronously only once. 

Are there any other solutions I am missing here?

Thanks a lot!

**Comment by jamesreggio** [comment]
Hi @reinvanimschoot — sorry for the slow response.

Your intuition is in line with how most React Native developers tend to initialize Apollo. Specifically, I tend to see people `await` the result of the init function inside `componentWillMount`, and then `setState` with the Apollo `client`, which gets passed into the `ApolloProvider` within the `render` function.

What I haven't seen before is a HOC like `withProvider` where you provide the Apollo Client in more than one place in your app. There's really no harm putting a single `ApolloProvider` at the very root of your app, since descendent components won't interact with it until you use a `graphql` HOC or query component. This is analogous to having a single Redux `Provider` at the root of your app, and then having individual `connect` HOCs where you need to access Redux state.

So, in summary, I'd suggest you use the `componentWillMount` approach in your app's root component, and drop the `withProvider` HOC. If that's not possible, you may have to store a global variable containing the `Promise` returned by your async init function, and then just `await` it from every `withProvider` instance.

I'm going to close the issue now, but I'm happy to continue the conversation — just let me know if anything wasn't clear.
